crates/iceberg/src/writer/mod.rs

```diff
@@ -15,14 +15,65 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! The iceberg writer module.
-
-use crate::spec::DataFileBuilder;
+//! Iceberg writer module.
+//!
+//! The writer API is designed to be extensible and flexible. Each writer is decoupled and can be create and config independently. User can:
+//! 1.Customize the writer using the writer trait.
+//! 2.Combine different writer to build a writer which have complex write logic.
+//!
+//! There are two kinds of writer:
+//! 1. FileWriter: Focus on writing record batch to different physical file format.(Such as parquet. orc)
+//! 2. IcebergWriter: Focus on the logical format of iceberg table. It will write the data using the FileWriter finally.
+//!
+//! # Simple example for data file writer:
+//! ```ignore
+//! // Create a parquet file writer builder. The parameter can get from table.
+//! let file_writer_builder = ParquetWriterBuilder::new(
+//!    0,
+//!    WriterProperties::builder().build(),
+//!    schema,
+//!    file_io.clone(),
+//!    loccation_gen,
+//!    file_name_gen,
+//! )
+//! // Create a data file writer using parquet file writer builder.
+//! let data_file_builder = DataFileBuilder::new(file_writer_builder);
+//! // Build the data file writer.
+//! let data_file_writer = data_file_builder.build().await.unwrap();
+//!
+//! data_file_writer.write(&record_batch).await.unwrap();
+//! let data_files = data_file_writer.flush().await.unwrap();
+//! ```
 
+pub mod base_writer;
 pub mod file_writer;
 
+use crate::{spec::DataFileBuilder, Result};
+use arrow_array::RecordBatch;
+
+type DefaultInput = RecordBatch;
 type DefaultOutput = Vec<DataFileBuilder>;
 
+/// The builder for iceberg writer.
+#[allow(async_fn_in_trait)]
+pub trait IcebergWriterBuilder<I = DefaultInput, O = DefaultOutput>:
+    Send + Clone + 'static
+{
+    /// The associated writer type.
+    type R: IcebergWriter<I, O>;
+    /// Build the iceberg writer.
+    async fn build(self) -> Result<Self::R>;
+}
+
+/// The iceberg writer used to write data to iceberg table.
+#[async_trait::async_trait]
```

Comment:
sad message: we can't use `async_fn_in_trait` here because it does not allow object safety so we can't construct something like `Box<dyn IcebergWriter>`. It's important to erase the type for complex writer.